underlay/src/main/java/bio/terra/tanagra/filterbuilder/impl/core/EntityGroupFilterBuilderBase.java

```diff
@@ -0,0 +1,364 @@
+                        underlay, notPrimaryEntity, selectedIds);
+                if (idSubFilter == null) {
+                  filtersForSingleEntityGroup.put(notPrimaryEntity, new ArrayList<>());
+                } else {
+                  filtersForSingleEntityGroup.put(
+                      notPrimaryEntity, new ArrayList<>(List.of(idSubFilter)));
+                }
+                EntityGroupFilterUtils.buildAllModifierFilters(
+                    underlay,
+                    List.of(notPrimaryEntity),
+                    criteriaSelector,
+                    valueData,
+                    modifiersSelectionData,
+                    filtersForSingleEntityGroup);
+                break;
+              default:
+                throw new SystemException(
+                    "Unsupported entity group type: " + entityGroup.getType());
+            }
+
+            List<EntityOutput> entityOutputsForSingleEntityGroup =
+                EntityGroupFilterUtils.mergeFiltersForDataFeature(
+                    filtersForSingleEntityGroup, BooleanAndOrFilter.LogicalOperator.AND);
+            entityOutputsForSingleEntityGroup.forEach(
+                entityOutput -> {
+                  List<EntityFilter> filters =
+                      filtersPerEntity.getOrDefault(entityOutput.getEntity(), new ArrayList<>());
+                  if (entityOutput.hasDataFeatureFilter()) {
+                    filters.add(entityOutput.getDataFeatureFilter());
+                  } else {
+                    filters = new ArrayList<>();
+                  }
+                  filtersPerEntity.put(entityOutput.getEntity(), filters);
+                });
+          });
+
+      // If there are multiple filters for a single entity, OR them together.
+      return EntityGroupFilterUtils.mergeFiltersForDataFeature(
+          filtersPerEntity, BooleanAndOrFilter.LogicalOperator.OR);
+    }
+  }
+
+  private Map<EntityGroup, List<Literal>> selectedIdsPerEntityGroup(
+      Underlay underlay, String serializedSelectionData) {
+    Map<EntityGroup, List<Literal>> selectedIdsPerEntityGroup = new HashMap<>();
+    Map<Literal, String> selectedIdsAndEntityGroups =
+        selectedIdsAndEntityGroups(serializedSelectionData);
+    selectedIdsAndEntityGroups.forEach(
+        (key, entityGroupId) -> {
+          EntityGroup entityGroup = underlay.getEntityGroup(entityGroupId);
+          List<Literal> selectedIds =
+              selectedIdsPerEntityGroup.containsKey(entityGroup)
+                  ? selectedIdsPerEntityGroup.get(entityGroup)
+                  : new ArrayList<>();
+          selectedIds.add(key);
+          selectedIdsPerEntityGroup.put(entityGroup, selectedIds);
+        });
+
+    selectedIdsPerEntityGroup.forEach(
+        (entityGroup, selectedIds) -> {
+          Collections.sort(selectedIds);
+        });
+    return selectedIdsPerEntityGroup;
+  }
+
+  private List<EntityGroup> selectedEntityGroups(
```

Comment:
nit: Could you add a comment here that this returns the list of entity groups that the user selected ids from, or the list of configured entity groups if the user selected no ids? I know that now from our discussions and reviewing this code, but seems like something I'm going to forget.

Reply:
Done.